engine/worker_task_conrrent_processor.go

```diff
@@ -0,0 +1,295 @@
+package engine
+
+import (
+	"context"
+	"fmt"
+	"github.com/xdblab/xdb-apis/goapi/xdbapi"
+	"github.com/xdblab/xdb/common/log"
+	"github.com/xdblab/xdb/common/log/tag"
+	"github.com/xdblab/xdb/common/ptr"
+	"github.com/xdblab/xdb/common/urlautofix"
+	"github.com/xdblab/xdb/config"
+	"github.com/xdblab/xdb/persistence"
+	"io/ioutil"
+	"net/http"
+	"time"
+)
+
+type workerTaskConcurrentProcessor struct {
+	rootCtx            context.Context
+	cfg                config.Config
+	taskToProcessChans chan persistence.WorkerTask
+	// for quickly checking if the shardId is being processed
+	currentShards map[int32]bool
+	// shardId to the channel
+	taskToCommitChans map[int32]chan<- persistence.WorkerTask
+	// shardId to the notifier
+	workerTaskNotifier map[int32]LocalNotifyNewWorkerTask
+	store              persistence.ProcessStore
+	logger             log.Logger
+}
+
+func NewWorkerTaskConcurrentProcessor(
+	ctx context.Context, cfg config.Config,
+	store persistence.ProcessStore, logger log.Logger) WorkerTaskProcessor {
+	bufferSize := cfg.AsyncService.WorkerTaskQueue.ProcessorBufferSize
+	return &workerTaskConcurrentProcessor{
+		rootCtx:            ctx,
+		cfg:                cfg,
+		taskToProcessChans: make(chan persistence.WorkerTask, bufferSize),
+		currentShards:      map[int32]bool{},
+		taskToCommitChans:  make(map[int32]chan<- persistence.WorkerTask),
+		workerTaskNotifier: make(map[int32]LocalNotifyNewWorkerTask),
+		store:              store,
+		logger:             logger,
+	}
+}
+
+func (w *workerTaskConcurrentProcessor) Stop(context.Context) error {
+	return nil
+}
+func (w *workerTaskConcurrentProcessor) GetTasksToProcessChan() chan<- persistence.WorkerTask {
+	return w.taskToProcessChans
+}
+
+func (w *workerTaskConcurrentProcessor) AddWorkerTaskQueue(
+	shardId int32, tasksToCommitChan chan<- persistence.WorkerTask, notifier LocalNotifyNewWorkerTask,
+) (alreadyExisted bool) {
+	exists := w.currentShards[shardId]
+	w.currentShards[shardId] = true
+	w.workerTaskNotifier[shardId] = notifier
+	w.taskToCommitChans[shardId] = tasksToCommitChan
+	return exists
+}
+
+func (w *workerTaskConcurrentProcessor) Start() error {
+	concurrency := w.cfg.AsyncService.WorkerTaskQueue.ProcessorConcurrency
+
+	for i := 0; i < concurrency; i++ {
+		go func() {
+			for {
+				select {
+				case <-w.rootCtx.Done():
+					return
+				case task, ok := <-w.taskToProcessChans:
```

Comment:
Question: IIUC, currently, the task in taskToProcessChans is executed sequentially, not in parallel?

Reply:
It will be in parallel because there are `concrrency` number of goroutines running to wait for the task. So the tasks will be distributed among all the goroutines and execute in paralllel